### ISSUE TITLE
Nebula name resolution for internal services

### ISSUE THREAD

**Issue by FDrebin** [slackhq/nebula#434]
Recently spun up Nebula on a few development hosts in our dev environment, we were successfully able to connect various databases and services across the Nebula network via their Nebula IP's which is awesome, but we also have some web services that terminate at load balancer and we use dns names for them, something like service.internal.ourdomain.com (not reachable via public internet) are unreachable or at least unreachable via its DNS name.

I read somewhere on one of the blogs/tutorials for Nebula setup that resolving internal DNS wasn't possible yet, but I did see in the example config for 1.30 that the lighthouse can act as DNS, is it possible to have the lighthouse act as DNS and be able to point clients of the Nebula network to something like service.internal.ourdomain.com on the Nebula range?

**Comment by caguiclajmg** [comment]
The internal DNS server has been out for quite some time, just set `serve_dns` in the config to true and the items under the `dns` key which is in the example config.

Lastly, you need to issue certificates with the correct `name` since the internal DNS server uses this for name resolution.

**Comment by FDrebin** [comment]
Hey thanks for the reply, so I will need to generate new certificates for service.internal.domain.com correct? And once that's done I can flip on the internal DNS and it should start being able to reach these services via Nebula?

**Comment by caguiclajmg** [comment]
Correct, you should be able to query the DNS server with that hostname if the certificate name is set up properly.

#110 might also be of interest

**Comment by FDrebin** [comment]
Hey thanks for that link, yea this is probably going to be trickier then I thought. Our load balancer servers up multiple DNS names for all our services, its really our single termination point for all our web services available.

I will have to poke around with the settings.

**Comment by FDrebin** [comment]
Hey, so following up with this. If I have multiple services reporting to the same load balancer, would I be able to have Nebula generate multiple DNS names, or have Nebula listen with multiple certificate names?

**Comment by caguiclajmg** [comment]
Unfortunately certificates can only have one name and certificate-node pairs are 1:1 so you can't set up multiple certificates per node as a workaround.

https://github.com/slackhq/nebula/blob/480036fbc8e50cef3649a98123706a6dd0072900/cert/cert.go#L36-L37

I was thinking about this the other way as well and needed the multiple name capability so I started to write a patch to support this. Fortunately, it seems the name isn't being used much anywhere except the firewall so this looks very feasible.

**Comment by FDrebin** [comment]
Yes that would be awesome!  I am not sure where I would even begin to try and pitch in with some of the "wants" that I have, I would love to be able to install Nebula on a server that provides DNS and simply use that as a one stop shop for resolving internal names and such, and it appears its a want for others as well. Some things are worth the wait.

**Comment by caguiclajmg** [comment]
If you're interested you can check out my fork with this feature [caguiclajmg/nebula:multi_cert_names](https://github.com/caguiclajmg/nebula/tree/multi_cert_names), just beware that it hasn't been extensively tested and is therefore not suitable for a production environment.

I've been running on a fleet of 20 machines (18 servers and 2 home devices) for about a day now and haven't seen any problems, the DNS resolution seems to be working as well, I can resolve multiple names that point to a single nebula IP.

**Comment by FDrebin** [comment]
> If you're interested you can check out my fork with this feature caguiclajmg/nebula:multi_cert_names,

I will give it a look at sometime this week or weekend, is that the lighthouse providing the DNS resolution via the DNS name for the certiticate correct? So I would just need to generate a certificate with the lighthouse CA for each DNS name I want that IP to respond to, such as` ./nebula-cert sign -name "service1.internal.domain.com, service2.internal.domain.com, portal.internal.domain.com" -ip "10.99.0.2/24"` or similar correct?

**Comment by caguiclajmg** [comment]
Yes, correct. Ideally, you would only need to re-issue the certificates with the intended names to use this (no configuration or changes).

That `nebula-cert` command is on point, just trim the extra spaces in between the commas.

**Comment by FDrebin** [comment]
And my (hopefully last question bothering you lol), I would only need to build from source for just the lighthouse correct, I wouldn't need to copy the newly built binaries to the other nodes?

**Comment by caguiclajmg** [comment]
> I wouldn't need to copy the newly built binaries to the other nodes?

Quite a PITA, but unfortunately you need to (i.e. generate new certificates for every node, install the binaries on every node). This fork introduces changes in the protobuf definition of the nebula certificate, without it the nodes would get confused on how to read the multiple names inside the certificate. I suggest having a backup strategy ready since this is basically a backward incompatible change to your setup.

> Edit : And for say iOS , would I need to go into the configuration of the hosts I have setup a define a public ip or dns name for that host, or is that just left blank and the lighthouse handles the translation if I go to a web browser and type in portal.internal.domain.co, ?

I have not used the iOS app so I can't say for sure, but if there is a way for you to change the device's DNS servers such that it points to the lighthouse serving the DNS records (warning: the nebula DNS server doesn't do recursive queries); then yes, you should be able to access a node via its internal names defined in the certificate.

**Comment by FDrebin** [comment]
So I finally fired this up, on iOS I get a certificate signature does not match with the ca.crt that i re-generated using the above pull from your repo. But it did successfully connect on a PC. I am trying now to see if the DNS stuff will work based on certificate name.

**Comment by JohnMaguire** [comment]
> So I finally fired this up, on iOS I get a certificate signature does not match with the ca.crt that i re-generated using the above pull from your repo. But it did successfully connect on a PC. I am trying now to see if the DNS stuff will work based on certificate name.

On mobile, I believe you will need to export the public key generated on iOS, then sign it using `nebula-cert sign -in-pub` and then copy the certificate back to the device (easiest via QR.)

The goal is to avoid moving the private key off of the mobile device itself.

**Comment by FDrebin** [comment]
Yup, did exactly that like I did when I was using the 1.3.0 build of Nebula, with the dev build from your repo I did the same, imported the ca.crt but it keeps flagging as not valid. Also went ahead and gave the Windows a whirl, for Windows 10 x64 I also receive 

`PS S:\Nebula> .\nebula.exe -config .\config.yml
time="2021-04-20T11:36:46-04:00" level=info msg="Firewall rule added" firewallRule="map[caName: caSha: direction:outgoing endPort:0 groups:[] host:any ip: proto:0 startPort:0]"
time="2021-04-20T11:36:46-04:00" level=info msg="Firewall rule added" firewallRule="map[caName: caSha: direction:incoming endPort:0 groups:[] host:any ip: proto:0 startPort:0]"
time="2021-04-20T11:36:46-04:00" level=info msg="Firewall started" firewallHash=21716b47a7a140e448077fe66c31b4b42f232e996818d7dd1c6c4991e066dbdb
time="2021-04-20T11:36:46-04:00" level=info msg="Main HostMap created" network=10.1.90.3/24 preferredRanges="[]"
time="2021-04-20T11:36:46-04:00" level=info msg="UDP hole punching enabled"
time="2021-04-20T11:36:46-04:00" level=info msg="Nebula interface is active" build=dev+20210420134215 interface=nebula1 network=10.1.90.3/24 udpAddr="[::]:4242"
time="2021-04-20T11:36:46-04:00" level=fatal msg="Activate failed: Failed to find the tap device in registry with specified ComponentId 'tap0901' and InterfaceName 'nebula1', TAP driver may be not installed or you may have specified an interface name that doesn't exist"
PS S:\Nebula>`

I also tried it on my known working version of Windows 10 where I was successful in using the windows TAP driver before with the 1.3.0 but was unsuccessful with it now. I am using the TAP/TUN driver found here : https://build.openvpn.net/downloads/releases/ , I am using tap-windows-9.24.2-I601-Win10.exe .

On linux (Ubuntu) it does establish a connection, so I am going to spin up a Ubuntu desktop and see if I can reach the host via its dns cert name.

**Comment by FDrebin** [comment]
Alrighty! Spun up a Ubuntu desktop, was able to get the nebula established to the lighthouse using the ca.crt. I am able to ping across the mesh, the one thing I can not do yet is work out the static map to point "nebula_IP" to ["dns_name"] , although, I believe the lgihthouse having the dns : true set should handle it.
